File: events/year_2023/day_02/day_02.py
"""
Advent of Code - Year 2023 - Day 2
https://adventofcode.com/2023/day/2
"""

import re


def preprocessing(puzzle_input: str) -> list[list[tuple[int, int]]]:
    """
    Parses the puzzle input string into a list of lists of tuples, each containing a number and the
    index of a color.
    """
    game = []
    for line in puzzle_input.splitlines():
        draws = [(int(n), "bgr".index(c)) for n, c in re.findall(r'([0-9]+)\s+(b|g|r)', line)]
        game.append(draws)
    return game


def solver(game: list[list[tuple[int, int]]]) -> tuple[int, int]:
    """
    Calculates and yields the sum of valid game IDs and the total power based on maximum values of
    each color in the provided game data.
    """
    power = 0
    sumid = 0
    for game_id, draws in enumerate(game, 1):
        if all(n + c < 15 for (n, c) in draws):
            sumid += game_id

        mb = max(n for (n, c) in draws if c == 0)
        mg = max(n for (n, c) in draws if c == 1)
        mr = max(n for (n, c) in draws if c == 2)
        power += mb * mg * mr

    return sumid, power

File: events/year_2023/day_02/day_02.rs
aoc::main!();

fn preprocessing(input: &str) -> Vec<Vec<(usize, usize)>> {
    input
        .lines()
        .map(|line| {
            Regex::new(r"([0-9]+)\s+(b|g|r)")
                .unwrap()
                .captures_iter(line)
                .map(|draw| draw.extract())
                .map(|(_, [n, c])| match c {
                    "b" => (n.parse().unwrap(), 0),
                    "g" => (n.parse().unwrap(), 1),
                    "r" => (n.parse().unwrap(), 2),
                    _ => unreachable!(),
                })
                .collect_vec()
        })
        .collect()
}

fn part_1(game: Vec<Vec<(usize, usize)>>) -> usize {
    game.iter()
        .enumerate()
        .filter(|(_, draws)| draws.iter().all(|(n, c)| n + c < 15))
        .map(|(n, _)| n + 1)
        .sum()
}

fn part_2(game: Vec<Vec<(usize, usize)>>) -> usize {
    game.iter()
        .map(|draws| {
            draws
                .iter()
                .fold((0, 0, 0), |(mb, mg, mr), (n, c)| match c {
                    0 => (mb.max(*n), mg, mr),
                    1 => (mb, mg.max(*n), mr),
                    2 => (mb, mg, mr.max(*n)),
                    _ => unreachable!(),
                })
        })
        .map(|(mb, mg, mr)| mb * mg * mr)
        .sum()
}
